diagnostics/diagnostics_manager.rs

```diff
@@ -0,0 +1,155 @@
+        } else {
+            None
+        };
+
+        Self { diagnostics, reporter, monitoring_server }
+    }
+
+    diagnostics_method! {
+        pub fn submit_database_metrics(&self, database_metrics: HashSet<DatabaseMetrics>);
+        pub fn submit_error(&self, database_name: Option<impl AsRef<str> + Hash>, error_code: String);
+        pub fn submit_action_success(&self, database_name: Option<impl AsRef<str> + Hash>, action_kind: ActionKind);
+        pub fn submit_action_fail(&self, database_name: Option<impl AsRef<str> + Hash>, action_kind: ActionKind);
+        pub fn increment_load_count(&self, database_name: &str, connection_: LoadKind);
+        pub fn decrement_load_count(&self, database_name: &str, connection_: LoadKind);
+    }
+
+    pub fn may_start_reporting(&self) {
+        if let Some(reporter) = &self.reporter {
+            reporter.may_start();
+        }
+    }
+
+    pub async fn may_start_monitoring(&self) {
+        if let Some(server) = &self.monitoring_server {
+            server.start_serving().await;
+        }
+    }
+}
+
+pub fn run_with_diagnostics<F, T>(
+    diagnostics_manager: &DiagnosticsManager,
+    database_name: Option<impl AsRef<str> + Hash>,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Result<T, Status>,
+{
+    let result = f();
+    submit_result_metrics(diagnostics_manager, database_name, action_kind, &result);
+    result
+}
+
+pub async fn run_with_diagnostics_async<F, Fut, T>(
+    diagnostics_manager: Arc<DiagnosticsManager>,
+    database_name: Option<impl AsRef<str> + Hash>,
+    action_kind: ActionKind,
+    f: F,
+) -> Result<T, Status>
+where
+    F: FnOnce() -> Fut,
+    Fut: Future<Output = Result<T, Status>> + Send,
+    T: Send,
+{
+    let result = f().await;
+    submit_result_metrics(&diagnostics_manager, database_name, action_kind, &result);
+    result
+}
+
+fn submit_result_metrics<T>(
+    diagnostics_manager: &DiagnosticsManager,
+    database_name: Option<impl AsRef<str> + Hash>,
+    action_kind: ActionKind,
+    result: &Result<T, Status>,
+) {
+    if !is_diagnostics_needed(database_name.as_ref()) {
+        return;
+    }
+
+    match result {
+        Ok(_) => diagnostics_manager.submit_action_success(database_name, action_kind),
+        Err(status) => {
+            diagnostics_manager.submit_action_fail(database_name.as_ref(), action_kind);
+            if let Some(error_code) = get_status_error_code(status) {
+                diagnostics_manager.submit_error(database_name, error_code.clone());
+            }
+        }
+    }
+}
+
+fn get_status_error_code(status: &Status) -> Option<String> {
+    if let Ok(details) = status.check_error_details() {
+        if let Some(error_info) = details.error_info() {
+            return Some(error_info.reason.clone());
+        }
+    }
+    None // status.code() can return too long string descriptions from external libraries
```

Comment:
I tried using `status.code()` as @dmitrii-ubskii suggested, but it doesn't seem correct. For example, in Authentication, we use these status codes, which only have the `Display` impls with very-very long strings. I'd assume that `error metrics` are exactly the metrics for our server's status codes, and the authentication issues can be tracked through `action metrics` and the amount of total/successful calls without error details. Didn't want to bother with this refactoring.

However, if you're okay with sending long descriptions of the errors in such cases, I can add them.

Reply:
yeah i think it's ok without the super long strings, we can imrpove this in the future if not illustrative enough!